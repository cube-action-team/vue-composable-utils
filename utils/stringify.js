const stringifyObject = (object, brace = ';') => {
  const keyValuePairs = Object.entries(object).filter(pair => pair[0] || pair[1]);

  return keyValuePairs.reduce((acc, item, index) => {
    if (index === keyValuePairs.length - 1) {
      return (acc = acc + `${item[0]}=${item[1]}`);
    }

    return (acc = acc + `${item[0]}=${item[1]}${brace}`);
  }, '');
};

export { stringifyObject };
