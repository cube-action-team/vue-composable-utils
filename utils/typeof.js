export const isClient = typeof window !== 'undefined';

export const isArray = Array.isArray;

export const isFunction = val => typeof val === 'function';

export const isString = val => typeof val === 'string';

export const isSymbol = val => typeof val === 'symbol';

export const isBoolean = val => typeof val === 'boolean';

export const isDate = val => isObject(val) && isFunction(val.getTime);

export const isNumber = val => typeof val === 'number';

export const isObject = val => val !== null && typeof val === 'object';

export const isElement = val => isObject(val) && !!val.tagName;

export function isPromise(val) {
  return isObject(val) && isFunction(val.then) && isFunction(val.catch);
}
