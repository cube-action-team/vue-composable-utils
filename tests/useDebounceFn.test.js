import { useDebounceFn } from '@/';
import { mount, createLocalVue } from '@vue/test-utils';
import { mountComposition } from './utils';
import VueCompositionAPI from '@vue/composition-api';

const localVue = createLocalVue();
localVue.use(VueCompositionAPI);

describe('useDebounceFn tests', () => {
  it('should be defined useDebounceFn', () => {
    expect(useDebounceFn).toBeDefined();
  });

  it('should calls fn method immediately if value of immediate is true', async () => {
    jest.useFakeTimers();

    const val = 1;
    const fn = jest.fn(value => value);
    const debouncedFn = useDebounceFn(
      {
        delay: 1000,
        immediate: true,
      },
      fn,
    );
    const result = await debouncedFn(val);

    expect(fn).toBeCalledWith(val);
    expect(result).toBe(val);
    expect(clearTimeout).toHaveBeenCalledTimes(0);
    expect(setTimeout).toHaveBeenCalledTimes(0);

    jest.useRealTimers();
  });

  it('should calls fn method after given delay', done => {
    expect.assertions(2);

    const val = 1;
    const fn = jest.fn(value => value);

    const debouncedFn = useDebounceFn(
      {
        delay: 1000,
        immediate: false,
      },
      fn,
    );

    debouncedFn(val).then(() => {
      expect(fn).toBeCalledTimes(1);
      expect(fn).toBeCalledWith(val);
      done();

      return null;
    });
  });

  it('should calls fn method only once within given delay', done => {
    expect.assertions(3);

    const val = 1;
    const val2 = 2;
    const fn = jest.fn(value => value);

    const wrapper = mount(
      mountComposition(() => {
        const debouncedFn = useDebounceFn(
          {
            delay: 1000,
            immediate: false,
          },
          fn,
        );
        return { debouncedFn };
      }),
      { localVue },
    );

    localVue.nextTick(() => {
      wrapper.vm.debouncedFn(val);
    });

    setTimeout(() => {
      wrapper.vm.debouncedFn(val2).then(() => {
        expect(fn).toBeCalledTimes(1);
        expect(fn).toBeCalledWith(val2);
        done();

        return null;
      });
    }, 500);

    expect(wrapper.html()).toMatchSnapshot();
  });
});
