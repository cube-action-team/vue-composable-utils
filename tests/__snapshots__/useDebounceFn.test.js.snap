// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDebounceFn tests should calls fn method only once within given delay 1`] = `<div></div>`;
